In [1]:
import pandas as pd
import numpy as np
from sklearn.preprocessing import StandardScaler
from sklearn.model_selection import train_test_split
from sklearn.tree import DecisionTreeClassifier
from sklearn.metrics import accuracy_score

In [2]:
from sklearn.datasets import load_breast_cancer
data = pd.DataFrame(load_breast_cancer()['data'],columns=load_breast_cancer()['feature_names']) 
data['y'] = load_breast_cancer()['target'] 

In [3]:
data

mean radius  mean texture  mean perimeter  mean area  mean smoothness  \
0          17.99         10.38          122.80     1001.0          0.11840   
1          20.57         17.77          132.90     1326.0          0.08474   
2          19.69         21.25          130.00     1203.0          0.10960   
3          11.42         20.38           77.58      386.1          0.14250   
4          20.29         14.34          135.10     1297.0          0.10030   
..           ...           ...             ...        ...              ...   
564        21.56         22.39          142.00     1479.0          0.11100   
565        20.13         28.25          131.20     1261.0          0.09780   
566        16.60         28.08          108.30      858.1          0.08455   
567        20.60         29.33          140.10     1265.0          0.11780   
568         7.76         24.54           47.92      181.0          0.05263   

     mean compactness  mean concavity  mean concave points  mean symmetry  \
0             0.27760         0.30010              0.14710         0.2419   
1             0.07864         0.08690              0.07017         0.1812   
2             0.15990         0.19740              0.12790         0.2069   
3             0.28390         0.24140              0.10520         0.2597   
4             0.13280         0.19800              0.10430         0.1809   
..                ...             ...                  ...            ...   
564           0.11590         0.24390              0.13890         0.1726   
565           0.10340         0.14400              0.09791         0.1752   
566           0.10230         0.09251              0.05302         0.1590   
567           0.27700         0.35140              0.15200         0.2397   
568           0.04362         0.00000              0.00000         0.1587   

     mean fractal dimension  ...  worst texture  worst perimeter  worst area  \
0                   0.07871  ...          17.33           184.60      2019.0   
1                   0.05667  ...          23.41           158.80      1956.0   
2                   0.05999  ...          25.53           152.50      1709.0   
3                   0.09744  ...          26.50            98.87       567.7   
4                   0.05883  ...          16.67           152.20      1575.0   
..                      ...  ...            ...              ...         ...   
564                 0.05623  ...          26.40           166.10      2027.0   
565                 0.05533  ...          38.25           155.00      1731.0   
566                 0.05648  ...          34.12           126.70      1124.0   
567                 0.07016  ...          39.42           184.60      1821.0   
568                 0.05884  ...          30.37            59.16       268.6   

     worst smoothness  worst compactness  worst concavity  \
0             0.16220            0.66560           0.7119   
1             0.12380            0.18660           0.2416   
2             0.14440            0.42450           0.4504   
3             0.20980            0.86630           0.6869   
4             0.13740            0.20500           0.4000   
..                ...                ...              ...   
564           0.14100            0.21130           0.4107   
565           0.11660            0.19220           0.3215   
566           0.11390            0.30940           0.3403   
567           0.16500            0.86810           0.9387   
568           0.08996            0.06444           0.0000   

     worst concave points  worst symmetry  worst fractal dimension  y  
0                  0.2654          0.4601                  0.11890  0  
1                  0.1860          0.2750                  0.08902  0  
2                  0.2430          0.3613                  0.08758  0  
3                  0.2575          0.6638                  0.17300  0  
4                  0.1625          0.2364                  0.07678  0  
..               

In [4]:
data.head()

mean radius  mean texture  mean perimeter  mean area  mean smoothness  \
0        17.99         10.38          122.80     1001.0          0.11840   
1        20.57         17.77          132.90     1326.0          0.08474   
2        19.69         21.25          130.00     1203.0          0.10960   
3        11.42         20.38           77.58      386.1          0.14250   
4        20.29         14.34          135.10     1297.0          0.10030   

   mean compactness  mean concavity  mean concave points  mean symmetry  \
0           0.27760          0.3001              0.14710         0.2419   
1           0.07864          0.0869              0.07017         0.1812   
2           0.15990          0.1974              0.12790         0.2069   
3           0.28390          0.2414              0.10520         0.2597   
4           0.13280          0.1980              0.10430         0.1809   

   mean fractal dimension  ...  worst texture  worst perimeter  worst area  \
0                 0.07871  ...          17.33           184.60      2019.0   
1                 0.05667  ...          23.41           158.80      1956.0   
2                 0.05999  ...          25.53           152.50      1709.0   
3                 0.09744  ...          26.50            98.87       567.7   
4                 0.05883  ...          16.67           152.20      1575.0   

   worst smoothness  worst compactness  worst concavity  worst concave points  \
0            0.1622             0.6656           0.7119                0.2654   
1            0.1238             0.1866           0.2416                0.1860   
2            0.1444             0.4245           0.4504                0.2430   
3            0.2098             0.8663           0.6869                0.2575   
4            0.1374             0.2050           0.4000                0.1625   

   worst symmetry  worst fractal dimension  y  
0          0.4601                  0.11890  0  
1          0.2750                  0.08902  0  
2          0.3613                  0.08758  0  
3          0.6638                  0.17300  0  
4          0.2364                  0.07678  0  

[5 rows x 31 columns]

In [5]:
data.shape

(569, 31)

In [6]:
data.isnull().sum()

mean radius                0
mean texture               0
mean perimeter             0
mean area                  0
mean smoothness            0
mean compactness           0
mean concavity             0
mean concave points        0
mean symmetry              0
mean fractal dimension     0
radius error               0
texture error              0
perimeter error            0
area error                 0
smoothness error           0
compactness error          0
concavity error            0
concave points error       0
symmetry error             0
fractal dimension error    0
worst radius               0
worst texture              0
worst perimeter            0
worst area                 0
worst smoothness           0
worst compactness          0
worst concavity            0
worst concave points       0
worst symmetry             0
worst fractal dimension    0
y                          0
dtype: int64

In [7]:
scaler = StandardScaler()
X = data.iloc[:, :-1]
y = data.iloc[:, -1]
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = .3, stratify = y, random_state = 50)

# Building Decision Tree Model

In [8]:
clf = DecisionTreeClassifier()
clf.fit(X_train, y_train)

DecisionTreeClassifier()

In [9]:
y_pred = clf.predict(X_test)

In [10]:
accuracy_rf = accuracy_score(y_test, y_pred)
accuracy_rf 

0.9181286549707602

### By changing hyper-parameters

In [62]:
parameters = {
    'criterion': ['gini', 'entropy'],
    'max_depth': [10, 20, 50, 100],
    'min_samples_leaf': [10, 20, 50]
}

In [63]:
clf = DecisionTreeClassifier()

In [64]:
from sklearn.model_selection import GridSearchCV
grid_search = GridSearchCV(clf, parameters)
grid_search.fit(X_train, y_train)

GridSearchCV(estimator=DecisionTreeClassifier(),
             param_grid={'criterion': ['gini', 'entropy'],
                         'max_depth': [10, 20, 50, 100],
                         'min_samples_leaf': [10, 20, 50]})

In [65]:
best_params = grid_search.best_params_
best_estimator = grid_search.best_estimator_

In [66]:
y_pred = best_estimator.predict(X_test)

In [67]:
accuracy_rf = accuracy_score(y_test, y_pred)
accuracy_rf 

0.9532163742690059

# Random Forest

In [17]:
from sklearn.ensemble import RandomForestClassifier

In [22]:
clf_rf = RandomForestClassifier()
clf_rf.fit(X_train, y_train)

RandomForestClassifier()

In [23]:
y_pred_rf = clf_rf.predict(X_test)

In [24]:
accuracy_rf = accuracy_score(y_test, y_pred_rf)
accuracy_rf 

0.9707602339181286

### By changing hyper-parameters

In [52]:
parameters = {
    'criterion': ['gini', 'entropy'],
    'max_depth': [10, 20, 50, 100],
    'min_samples_leaf': [10, 20, 50]
}

In [53]:
clf_rf = RandomForestClassifier()

In [58]:
from sklearn.model_selection import GridSearchCV
grid_search_rf = GridSearchCV(clf_rf, parameters)
grid_search_rf.fit(X_train, y_train)

GridSearchCV(estimator=RandomForestClassifier(),
             param_grid={'criterion': ['gini', 'entropy'],
                         'max_depth': [10, 20, 50, 100],
                         'min_samples_leaf': [10, 20, 50]})

In [59]:
best_params = grid_search_rf.best_params_
best_estimator = grid_search_rf.best_estimator_

In [60]:
y_pred = best_estimator.predict(X_test)

In [61]:
accuracy_rf = accuracy_score(y_test, y_pred)
accuracy_rf 

0.9649122807017544

# Gradient Boosting Method

In [25]:
from sklearn.ensemble import GradientBoostingClassifier

In [28]:
clf_gb = GradientBoostingClassifier()
clf_gb.fit(X_train, y_train)

GradientBoostingClassifier()

In [30]:
y_pred_gb = clf_gb.predict(X_test)

In [31]:
accuracy_gb = accuracy_score(y_test, y_pred_gb)
accuracy_gb

0.9590643274853801

# K-Nearest Neighbour

In [32]:
from sklearn.neighbors import KNeighborsClassifier

In [34]:
clf_knn = KNeighborsClassifier()
clf_knn.fit(X_train, y_train)

KNeighborsClassifier()

In [35]:
y_pred_knn = clf_knn.predict(X_test)

In [36]:
accuracy_knn = accuracy_score(y_test, y_pred_knn)
accuracy_knn

0.9239766081871345

# From above Classification Models :
###    Accuracy of Decision tree:  0.918
###    Accuracy of Decision tree with Hyper-Parameters:  0.953
###    Accuracy of Random Forest: 0.970  
###    Accuracy of Random Forest with Hyper-Parameters: 0.964
###    Accuracy of Gradient Boosting : 0.959
###    Accuracy of K-Nearest Neighbour:  0.923




    
    

# By seeeing above accuracies we can say that random Forest classification model performing best then other models , where it is giving 97 % accuracy.

# END